.github/workflows/formatting.yml

```diff
@@ -0,0 +1,103 @@
+name: formatting-check
+
+on:
+  push:
+    # Only run on pushes to pull request branches
+    branches-ignore:
+      - 'master'
+      - 'master-*'
+  pull_request:
+
+jobs:
+
+  # Checking for correct formatting of branch for C code changes
+  check-formatting:
+    name: Formatting Check (clang 9)
+    runs-on: ubuntu-18.04
+    container: ubuntu:18.04
+    steps:
+
+      # Cache Rust stuff.
+      - name: Cache cargo registry
+        uses: actions/cache@v1
+        with:
+          path: ~/.cargo/registry
+          key: cargo-registry
+
+      - name: Install dependencies
+        run: |
+          apt update
+          apt -y install \
+                libpcre3 \
+                libpcre3-dev \
+                build-essential \
+                autoconf \
+                automake \
+                git \
+                libtool \
+                libpcap-dev \
+                libnet1-dev \
+                libyaml-0-2 \
+                libyaml-dev \
+                libcap-ng-dev \
+                libcap-ng0 \
+                libmagic-dev \
+                libnetfilter-queue-dev \
+                libnetfilter-queue1 \
+                libnfnetlink-dev \
+                libnfnetlink0 \
+                libhiredis-dev \
+                libjansson-dev \
+                libpython2.7 \
+                make \
+                python \
+                rustc \
+                software-properties-common \
+                wget \
+                zlib1g \
+                zlib1g-dev
+      - name: Install packages for clang-format 9
+        run: |
+          wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
+          # no need to install full clang using https://apt.llvm.org/llvm.sh
+          # using clang 9
+          add-apt-repository "deb http://apt.llvm.org/bionic/ llvm-toolchain-bionic-9 main"
+          apt-get update
+          apt-get install -y clang-format-9
+      - name: Install cbindgen
+        run: cargo install --force --debug --version 0.14.1 cbindgen
+      - run: echo "::add-path::$HOME/.cargo/bin"
+      # If master has newer commits, github adds a merge commit from master into
```

Comment:
That was a major pain-in -the-butt. Hence I also tested with branches off old master.

Reply:
And it seems to break on this repo, but works in my private repo. Argh. Unless the current formatting-check's checkout failure is not related to this...

Reply:
Nah, the problem is how I fixed the "don't give me the bloody merge commit if master is newer, I want to only have branch commits". Seems this will not work on the OISF repo as the OISF repo does not know about my branch in my fork. Makes sense.

Back to square one.

@jasonish if you have any idea how to do that in github action correctly, that'll be great. See comments to the checkout action for what the problem is. Tried a multitude of things among them the current:

```
      - name: Checkout HEAD on branch, not merge commit!
        uses: actions/checkout@v1
        with:
          ref: ${{ github.head_ref }}
#          ref: ${{ github.event.pull_request.head.sha }}
```
The commented one did not work as I think it referred to the merge request.

I also played around with `actions/checkout@v2` but that did not work due some git mismatch and it downloaded the code as tgz or so instead of cloning it.

I know I can solve it by manually peeling off the merge commit if it exists in the run command, but I'd rather use something that works with the github checkout action if popssible...

Reply:
I'm still unclear even after reading the comment.  With a plain:
```
uses: actions/checkout@v1
```
are you saying that a merge commit is created if master is newer? I don't believe that is the case. It would help if the comment explained why the `ref` is used all I suppose.

Reply:
Sorry for the unclear comments.

Yep, exactly.

github does in fact add a merge if master is newer. See e.g. this test branch where I print the newest two commits after checking out the branch with ``actions/checkout@v1``: https://github.com/OISF/suricata/runs/875169737?check_suite_focus=true 

E.g.
```
Run git log --oneline -2
c1b491ecb Merge 0c804f44a42ee162160e34ceab3f0df77a290097 into 98318393888909293edcfc22f0158de4350182f7
0c804f44a TEST: action: use of plain action results in merge commit
```

``0c804f44a`` is the last commit on the branch. github added a merge ``c1b491ecb`` with the changes from master. It all makes sense for normal builds. Just does not make sense for anything that only wants to verify things on the branch itself. 

It seems to be a known behaviour from my googling and all work-arounds so far don't work as we're merging from a fork's branch. The work-arounds by specifying ``with/ref`` seem to only work on the repo where the branch exists.

Reply:
Very interesting.  The workflow on "push", checks out at that specific commit, getting behaviour you expect.  However, the "pull_request" workflow results in this merge...

Is it necessarily a bad though? We somewhat expect pull requests to be made against the tip of the branch they are made against, so in most cases its a non-issue.  And as time goes, and a good back log of commits are formatted OK, these warnings should go away right?

Reply:
Nice find! I've never looked into that. The other one is that every push to an existing pull request is a ``push`` event in your private repo, but a ``pull_request`` event in the OISF repo.

Anyhow, back to your question...
It's not _that_ bad, but I find it substandard from a usability point-of-view for developers. Why show "wrong formatting" of other code? Better be ready for the support questions in the forums about that.

Use case: The developer branched off master last week (or maybe yesterday) for her feature/bug and new commits have been pushed to master in the OISF repo since then. She's done developing and testing her feature/bug so  today she's opening a pull request and sees "wrong formatting" not in her code, but in some other files she has not touched (as the latest commits on master are not properly formatted). That sucks.

While it's recommended to rebase your branch on the absolutely latest master just before opening a pull request for the OISF repo, I think that puts an unnecessary burden on the developer. In addition, there's also an albeit very small chance that while the developer is opening her pull request, another pull request has been merged into master.

Having said that, yes, in an ideal world, it'll go away over time as all the commits have the proper formatting. But then I highly recommend having the formatting check fail and make it a gating requirement for merges instead of just a warning.

Anyhow, we can avoid all the developer questions about "seeing wrong formatting for someone else's code" in weird and not so weird edge cases by simply ignoring the "github merge commit" in the action. I find it a small preventative step to avoid future work and questions. Btw, I have an implementation that does that.

Reply:
Fix will be in v5 that ignores the "merge commit" if there is one, i.e. for pull requests.